addons/l10n_in_edi/models/einvoice_api.py

```diff
@@ -0,0 +1,100 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from datetime import timedelta
+
+from odoo import fields, models, api, _
+from odoo.exceptions import AccessError
+
+from odoo.addons.iap import jsonrpc
+
+import logging
+
+_logger = logging.getLogger(__name__)
+
+DEFAULT_IAP_ENDPOINT = "https://l10n-in-edi.api.odoo.com"
+DEFAULT_IAP_TEST_ENDPOINT = "https://l10n-in-edi-proxy-demo.odoo.com"
+
+
+class L10nInEdiWebService(models.AbstractModel):
+    _name = "l10n.in.edi.web.service"
+    _description = "E-invoice API (IN)"
+
+    @api.model
+    def _get_token(self, company):
+        if company._l10n_in_edi_token_is_valid():
+            return company.l10n_in_edi_token
+        elif company.l10n_in_edi_username and company.sudo().l10n_in_edi_password:
+            self._authenticate(company)
+            return company.l10n_in_edi_token
+        return False
+
+    @api.model
+    def _connect_to_server(self, company, url_path, params):
+        user_token = self.env["iap.account"].get("l10n_in_edi")
+        params.update(
+            {
+                "account_token": user_token.account_token,
+                "dbuuid": self.env["ir.config_parameter"].sudo().get_param("database.uuid"),
+                "username": company.l10n_in_edi_username,
+                "gstin": company.vat,
+            }
+        )
+        if company.l10n_in_edi_test_env:
+            default_endpoint = DEFAULT_IAP_TEST_ENDPOINT
+        else:
+            default_endpoint = DEFAULT_IAP_ENDPOINT
+        endpoint = self.env["ir.config_parameter"].sudo().get_param("l10n_in_edi.endpoint", default_endpoint)
+        url = "%s%s" % (endpoint, url_path)
+        try:
+            return jsonrpc(url, params=params, timeout=25)
+        except AccessError as e:
+            _logger.warning("Connection error: %s", e.args[0])
+            return {
+                "error": [
+                    {
+                        "code": "404",
```

Comment:
Why code 404?  Because it is a known one?

Reply:
yes

Reply:
also, we set blocking_level to the warning if code is 404